Desktop notifications blog post (#248)
shipshapecode/website-nuxt: content/blog/posts/2022-08-04-add-desktop-notifications.md

---
authorId: chuckcarpenter
categories:
  - notifications
  - javascript
date: ''
slug: add-desktop-notifications
title: Adding desktop notifications to your app/site
---

There are times when your application needs to inform users to act on something more

> timely than a normal notification (often known as toast notifications) may allow.

Reviewer: ```suggestion
timely than a normal notification (often known as toast notifications) allows.
```

If you need a user to do something and can't wait until they return to the application for them to notice, or just want to engage them when something new is available, desktop notifications might be a good fit for your app.

> Desktop notifications are supported in all modern browsers (sorry Internet Explorer, you're not invited). However, the user will have to opt in and they will also have to allow notifications in their OS settings. One they have opted in, you can notify them on the system level rather than just within your web browser page/tab.

Reviewer: ```suggestion
Desktop notifications are supported in all modern browsers (sorry Internet Explorer, you're not invited). However, the user will have to opt-in and they will also have to allow notifications in their OS settings. Once they have opted in, you can notify them on the system level rather than just within your web browser page/tab.
```

## What we'll work on today

> Let's create a basic notification that checks for and then gets the user's permission. Then we can show a notification about our cool new blog posts. We can then add some actions to allow the user to opt it future notifications as to when new posts are published. There's more to it, but we'll get to the nuances shortly.

Reviewer: ```suggestion
Let's create a basic notification that checks for, and then gets the user's permission. Then we can show a notification about our cool new blog posts. We can then add some actions to allow the user to opt-in for future notifications as to when new posts are published. There's more to it, but we'll get to the nuances shortly.
```

For this I'm going to create a singleton pattern file since I need it to load and run when loaded. I find this pattern lends itself well to our needs and is the most readable.

```js
class NotificationService {
  constructor() {
    // whatever you want once instance is created
  }

  async _isShowNotificationPermitted() {
    // ...
  }

  async _showNotification() {
    // ...
  }
}

new NotificationService();
```

Using this javascript class I am adding a couple methods to handle checking the permission settings, requesting permissions, and then showing a notification if granted. Let's start with checking for and asking for permission to show notifications from our site.  

```js
async _isShowNotificationPermitted() {
  if ('Notification' in window) {
    let { permission } = Notification;
    if (permission !== 'granted') {
      permission =  await Notification.requestPermission();
    }

>     if (permission !== 'denied') {

Reviewer 2: This is probably not needed because I'm not sure how it would come back as `default` here, but a belt and suspenders approach might also cover `default` here so all possible values are handled.

```suggestion
    if (permission !== 'denied' || permission !== 'default') {
```

      return true;
    }
  }

  return false;
}
```

> Let's break down this method a bit. First we're checking for support in the browser, in the off chance your user doesn't have a supported browser, otherwise why have those needless errors? The default value from `Notification.permission` is just that, `default`, which we treat essentially the same as `denied`. In that case we trigger `requestPermission()` in the API to request access from the user. If the user approves, we can move on to showing them notifications until this permission is revoked. Let's look then at a method for showing one.

Reviewer: ```suggestion
Let's break down this method a bit. First, we're checking for support in the browser, in the off chance your user doesn't have a supported browser, otherwise why have those needless errors? The default value from `Notification.permission` is just that, `default`, which we treat essentially the same as `denied`. In that case, we trigger `requestPermission()` in the API to request access from the user. If the user approves, we can move on to showing them notifications until this permission is revoked. Let's look then at a method for showing one.
```

```js
async _showNotification() {
  const permission = await this._isShowNotificationPermitted();

  if (permission !== 'denied') {
    new Notification('Be notified of new posts', {
      body: 'Let us notifiy you of new posts on our site!',
    });
  }
}
```

This one is a bit more straightforward. We check permission, as noted in the previous code example, and as long as we're not denied, a notification can be shown. There's really only one required argument, a title. Here we are adding a basic option for body text as well. This will just show a notification and the user can dismiss it or select to go to browser settings, if they'd like to change the permissions. Let's make sure our instance fires off when loaded before we move on.

```js
class NotificationService {
  constructor() {
    this._showNotification();
  }

// ...the rest of the class
```

![Basic desktop notification](/img/blog/add-desktop-notifications/basic-notification.png)

> This gets us a basic notification, but without any interaction from the user. To do that, we have to add actions to the options hash that you can watch and respond too. The spec requires that we only can do that if we're triggering the notification via a service worker and responding to events there. Let's create a basic service worker file and add a listener. Create a file called `service-worker.js` and add the following code.

Reviewer: ```suggestion
This gets us a basic notification, but without any interaction from the user. To do that, we have to add actions to the options hash that you can watch and respond to. The spec requires that we only can do that if we're triggering the notification via a service worker and responding to events there. Let's create a basic service worker file and add a listener. Create a file called `service-worker.js` and add the following code.
```

not sure if this was meant to be `too` as in also or `to`

> This gets us a basic notification, but without any interaction from the user. To do that, we have to add actions to the options hash that you can watch and respond too. The spec requires that we only can do that if we're triggering the notification via a service worker and responding to events there. Let's create a basic service worker file and add a listener. Create a file called `service-worker.js` and add the following code.

Reviewer 3: ```suggestion
## Add title about adding user interaction

So far we have a basic notification; one without any interaction with the user. To add interactivity, we have to add `actions` to the options hash, which allows us to watch for and respond to the users actions.

The spec states that in order to add actions, we must trigger the notification via a service worker as well as respond to events there. Let's create a basic service worker file called `service-worker.js` and add the following event listener:
```

what is the spec? link could be helpful

```js
self.addEventListener('notificationclick', function(event) {
  const clickedNotification = event.notification;
  clickedNotification.close();

  // Do something as the result of the notification click
  const promiseChain = doSomething();
  event.waitUntil(promiseChain);
});
```

> In this we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.

Reviewer: ```suggestion
In this snippet we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.
```

Reviewer: stylistic choice but naming what we are referring comes off more direct and clear

> In this we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.

Reviewer 3: ```suggestion
In this event listener, we're watching for a notification click that tells us a user interacted has occurred. We'll want to register our worker in the notification service and have it trigger the following notification flow.
```

```js
_startServiceWorker(notificationService) {
  if ('serviceWorker' in navigator) {
    navigator.serviceWorker.register('service-worker.js')
    .then(function(registration) {
      registration.addEventListener('updatefound', function() {
        // If updatefound is fired, it means that there's
        // a new service worker being installed.
        const installingWorker = registration.installing;
        console.log('A new service worker is being installed:',
          installingWorker);

        // You can listen for changes to the installing service worker's
        // state via installingWorker.onstatechange
      });
      
      notificationService._showNotification(registration);
    })
    .catch(function(error) {
      console.log('Service worker registration failed:', error);
    });
  } else {
    console.log('Service workers are not supported.');
  }
}
```

In the above, we check for service worker support and then register our newly created worker file. After it's installled we show the notification in an updated `_showNotification` method. 

```js
async _showNotification(worker) {
  const permission = await this._isShowNotificationPermitted();

>   if (permission !== 'denied') {

Reviewer 3: It can be confusing to have most of the directions in paragraph form and some as commented out code - it might get missed. Maybe move those commented out lines out and refer to lines A, B, C...

    worker.showNotification('Be notified of new posts', {
      body: 'Let us notifiy you of new posts on our site!',
      requireInteraction: true,
      actions: [
        {
          action: 'dismiss',
          title: 'Dismiss'
        },
        {
          action: 'subscribe',
          title: 'Subscribe'
        }
      ]
    });
  }
}
```

> The difference now is that we can set actions on the notification options and allow the user to select something from that notification. This click will get passed to the worker and given what the action is, we can react and do something. So, let's do just that and act on the user's response, which we're requiring with `requireInteraction`.

Reviewer 3: ```suggestion
These changes allow us to set actions on the notification options and allow the user to interact with the notification, in this case by clicking [what they're clicking]. This click will get passed to the worker so we can react with a response. Let's do just that and act on the user's response, with `requireInteraction`.
```

- maybe show the before and after together so the change is more visible/memorable
- not sure if the suggested change is true to your original meaning

```js
self.addEventListener('notificationclick', function(event) {
  const { action, notification } = event;
  notification.close();
  
  if (action === 'subscribe') {
    const promiseChain = clients.openWindow('https://feedly.com/i/subscription/feed%2Fhttps%3A%2F%2Fshipshape.io%2Ffeed.xml');
    event.waitUntil(promiseChain);
  }
});
```

> Here I've updated the worker to check for an action if the user click's `subscribe` button in the notification, when it's captured and checked we open a new window with out RSS subscription link. If they click `dismiss`, nothing happens and the notification is dismissed. In case you want some kind of persistence, we could add some way to track the user selecting to dismiss and save that state to something like localStorage.

Reviewer: ```suggestion
Here I've updated the worker to check for an action if the user click's `subscribe` button in the notification, when it's captured and checked we open a new window without RSS subscription link. If they click `dismiss`, nothing happens and the notification is dismissed. In case you want some kind of persistence, we could add some way to track the user selecting to dismiss and save that state to something like localStorage.
```

> Here I've updated the worker to check for an action if the user click's `subscribe` button in the notification, when it's captured and checked we open a new window with out RSS subscription link. If they click `dismiss`, nothing happens and the notification is dismissed. In case you want some kind of persistence, we could add some way to track the user selecting to dismiss and save that state to something like localStorage.

Reviewer 3: ```suggestion
Here I've updated the worker to check if the user clicks the `subscribe` button. When **it's** captured and checked, we open a new window with out RSS subscription link. 

If they click `dismiss`, nothing happens and the notification is dismissed. If you want to add persistence, you could track the user's decision to dismiss and save that state to something like `localStorage`.
```

- what is the IT being captured? the click? what is checked for?
- with out? with our?
- for bonus seo - include the RSS as a link if it's possible

![Desktop notification with actions](/img/blog/add-desktop-notifications/notification-with-actions.png)

> 

Reviewer 3: ## Add a sub header here 

maybe something like, "Our final product: [what it does]"

> In the end, we have a class to setup our service worker and show a desktop notification to the user to allow them to subscribe to our blog posts. Here's what it looks like all together. 

Reviewer 3: ```suggestion
In summary, we have created a class to set up a service worker which shows the user a desktop notification that asks if they'd like to subscribe to our blog posts. 

Here's what it looks like all together:
```

```js
class NotificationService {
  constructor() {
    this._startServiceWorker(this);
  }

  _startServiceWorker(notificationService) {
    if ('serviceWorker' in navigator) {
      navigator.serviceWorker.register('/landing/js/service-worker.js')
      .then(function(registration) {
        registration.addEventListener('updatefound', function() {
          const installingWorker = registration.installing;
          console.log('A new service worker is being installed:', installingWorker);
        });

        notificationService._showNotification(registration);
      })
      .catch(function(error) {
        console.log('Service worker registration failed:', error);
      });
    } else {
      console.log('Service workers are not supported.');
    }
  }

  async _isShowNotificationPermitted() {
    if ('Notification' in window) {
      let { permission } = Notification;

>       if (permission !== 'granted') {

Reviewer 2: Will this re-request the permission if it's already been denied?

Author: ah yes, so this would be better

```
if (permission === 'granted') {
        return true;
      }
      if (permission !== 'denied') {
        permission =  await Notification.requestPermission();
      }
```

        permission =  await Notification.requestPermission();
      }
      if (permission !== 'denied') {
        return true;
      }
    }

    return false;
  }

  async _showNotification(worker) {
    const permission = await this._isShowNotificationPermitted();
    if (permission !== 'denied') {
      worker.showNotification('Be notified of new posts', {
      body: 'Let us notifiy you of new posts on our site!',
        actions: [
          {
            action: 'dismiss',
            title: 'Dismiss'
          },
          {
            action: 'subscribe',
            title: 'Subscribe'
          }
        ]
      });
    }
  }
}

const notifications = new NotificationService();
Object.freeze(notifications);
```

> ## In closing

Reviewer 3: suggest moving this header up to line 170

Desktop notifications are supported in all _modern_ browsers (sorry Internet Explorer, you're not invited). Keep in mind the user will have to opt-in and they will also have to allow notifications in their OS settings in order to receive notifications on the system level rather than only within their web browser. I will be covering both these steps in this tutorial.

## What we'll work on today: setting up a basic desktop notification 

Let's create a basic feature that checks for, and gets the user's permission so we can notify them about our cool new blog posts. We can add actions so the user can opt-in for future notifications when new posts are published as well. 

For this we're going to create a singleton pattern file since it's needed to load and run when the site is initially loaded and won't change after or be extended in the future, and is the most human readable.

```js
class NotificationService {
  constructor() {
    // whatever is required when new instance created, completed later
  }

  async _isShowNotificationPermitted() {
    // added later in the post
  }

  async _showNotification() {
    // added later in the post
  }
}

new NotificationService();
```

> We arecreating the JavaScript class, `NotificationService`, with methods that will handle checking the permission settings, requesting permissions, and then showing a notification if granted. 

Reviewer 4: arecreating -> are creating
